### DISCUSSION TITLE
staircase procedure

### DISCUSSION THREAD

**Discussion by dokato** [jspsych/jsPsych#735] [Q&A] [answered]
It's not clear to me how to implement a staircase procedure with jsPsych. I know that I can use loop, like this.

```js
var block = {
    type: "RDK",
    choices: ['z', 'm'],
    correct_choice: "z",
    coherence: 0.6,
    coherent_direction: [0],
};

var loop_node = {
    timeline: [block],
    loop_function: function(data){
      if (data.values()[0].correct === true)
        return false;
      console.log(data.values()[0].correct);
      return true;
    }
}

timeline.push(loop_node);
```
Inside the loop function I would set my condition for convergence, but how to change the content of block? In particular coherence value?

**Comment by jodeleeuw** [comment, marked as answer]
You can set the coherence parameter to be a function that returns the current value of the staircase:

```js
var block = {
    type: "RDK",
    choices: ['z', 'm'],
    correct_choice: "z",
    coherence: function(){
	return staircase_coherence;
    },
    coherent_direction: [0],
};
```

If you define this value globally then you can update it in the `loop_function`.

If you want to avoid a global variable then you can compute the coherence directly in the function that returns it.

**Reply by dokato** [reply]
Great, that works. Thanks for rapid response.

BTW there's one more thing to it, in this case RDK must be obviously called without time limit:

```js
var block = {
    type: "RDK",
    choices: ['z', 'm'],
    correct_choice: "z",
    coherence: function(){
	return staircase_coherence;
    },
    coherent_direction: [0],
   trial_duration: -1
};
```

**Comment by becky-gilbert** [comment]
Hi @dokato, I just wanted to jump in to let you know about some JavaScript libraries for staircase procedures, in case you're interested. I've used both of these with jsPsych in the same way that Josh describes (i.e. using functions as parameters, where the function returns the difficulty level for the current trial).
https://github.com/hadrienj/StaircaseJS
https://github.com/becky-gilbert/jsStaircase
Both of these libraries will calculate the difficulty level for the next trial, based on the participant's performance and a number of parameters like the up/down rule, step size, limits on the difficulty and/or step size values, etc.
StaircaseJS is nicely documented with examples. My version, jsStaircase, is similar to StaircaseJS but uses a different staircase method (allows for dynamic step sizes) and includes some different functions and parameters.

**Reply by dokato** [reply]
Much appreciated, @becky-gilbert. It's really useful.
